tests/wol/test_wol.py

```diff
@@ -55,495 +41,377 @@ def m2b(mac: str) -> bytes:
-    random_dut_port, random_ptf_port = random.choice(connected_dut_intf_to_ptf_index)
-    logging.info("Test with random dut port %s and ptf port index %s" % (random_dut_port, random_ptf_port))
-
-    def validate_wol_packets(pkts):
-        pytest_assert(len(pkts) == 1, "Unexpected pkts count %s" % len(pkts))
-        pkt = pkts[0]
-        pytest_assert(pkt.dst == target_mac, "Unexpected dst mac %s" % pkt.dst)
-        pytest_assert(pkt.src == dut_mac, "Unexpected src mac %s" % pkt.src)
-        pytest_assert(pkt.type == ETHER_TYPE_WOL_DEC)
-        pytest_assert(pkt.load == build_magic_packet_payload(target_mac))
-
-    with capture_and_check_packet_on_dut(
-        duthost=ptfhost,
-        interface='eth'+str(random_ptf_port),
-        pkts_filter=WOL_ETHER_PKT_FILTER,
-        pkts_validator=validate_wol_packets
+def get_packets_on_specified_ports(ptfadapter, verifier=None, ports=None, device_number=0, duration=1, timeout=None):
+    """
+    Get the packets on the specified ports and device for the specified duration
+    """
+    logging.info("Get pkts on device %d, port %r", device_number, ports)
+
+    received_pkts_res = {}
+    start_time = time.time()
+    while (time.time() - start_time) < duration:
+        result = testutils.dp_poll(ptfadapter, device_number=device_number, timeout=timeout)
+        logging.info(result)
+        if isinstance(result, ptfadapter.dataplane.PollSuccess) and (ports is None or result.port in ports):
+            if verifier is None or verifier(result.packet):
+                if result.port in received_pkts_res:
+                    received_pkts_res[result.port].append(result)
+                else:
+                    received_pkts_res[result.port] = [result]
+    return received_pkts_res
+
+
+def verify_packet(ptfadapter, verifier, port, count=1, interval=None, device_number=0, duration=1, timeout=None):
+    verify_packets(ptfadapter, verifier, [port], count, interval, device_number, duration, timeout)
+
+
+def verify_packets(ptfadapter, verifier, ports, count=1, interval=None, device_number=0, duration=1, timeout=None):
+    received_pkts = get_packets_on_specified_ports(ptfadapter, verifier, None, device_number, duration, timeout)
+    pytest_assert(set(received_pkts.keys()) == (set(ports) if count != 0 else set()),
+                  "Received packets on ports other than {}: {}".format(ports, list(received_pkts.keys())))
+    pytest_assert(all(map(lambda pkts: len(pkts) == count, received_pkts.values())),
+                  "Did not receive exactly {} of expected packets on all {}: {}".format(count, ports, received_pkts))
+    if count >= 2 and interval is not None:
+        for results in received_pkts.values():
+            ts = list(map(lambda result: result.time, results))
+            ts_diff = [ts[i] - ts[i - 1] for i in range(1, len(ts))]
+            pytest_assert(all(map(lambda diff: abs(diff * 1000 - interval) < 100, ts_diff)),
```

Comment:
no, it's not that accurate

Reply:
Hi @Xichen96, in the physical topo, the wol packet goes this path: Mx -> LeafFanout -> RootFanout -> Server (PTF). There is only 3 jumps in a single cluster. The packet transmission delay should be less than 1ms in a cluster, so we set the tolerant interval error as 5ms in testcase.
Can you please run this test and check if it can pass with 5ms? If 5ms not enough, what's the minimum value to pass the testcase. If the error is too high (like more than 10ms), we need to investigate which part introduced the interval error. Is there any a Rust library issue or fanout forwarding issue...